File: src/2685.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        retVal = 0

        adjacency = [[] for _ in range(n)]
        for u, v in edges:
            adjacency[u].append(v)
            adjacency[v].append(u)

        visited = [False] * n

        for vertex in range(n):
            if visited[vertex] == True:
                continue
            visited[vertex] = True

            # BFS to find all vertices in the current component
            component = []
            bfsQueue = [vertex]
            while bfsQueue:
                current = bfsQueue.pop(0)
                component.append(current)

                # Process neighbors
                for neighbor in adjacency[current]:
                    if visited[neighbor] == True:
                        continue
                    bfsQueue.append(neighbor)
                    visited[neighbor] = True

            # Check if component is complete (all vertices have the right number of edges)
            isComplete = True
            for node in component:
                componentSize = len(component)
                adjacencyNodeSize = len(adjacency[node])
                if adjacencyNodeSize != componentSize - 1:
                    isComplete = False
                    break
            if isComplete == True:
                retVal += 1

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for n, edges in zip([6, 6], [[[0, 1], [0, 2], [1, 2], [3, 4]], [[0, 1], [0, 2], [1, 2], [3, 4], [3, 5]]]):
            # /* Example
            #  *  Input: n = 6, edges = [[0,1],[0,2],[1,2],[3,4]]
            #  *  Output: 3
            #  *
            #  *  Input: n = 6, edges = [[0,1],[0,2],[1,2],[3,4],[3,5]]
            #  *  Output: 1
            #  */
            logging.info("Input: n = %d, edges = %s", n, edges)

            retVal = pSolution.countCompleteComponents(n, edges)
            logging.info("Output: %d", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)

File: src/2685.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int countCompleteComponents(int n, int** edges, int edgesSize, int* edgesColSize) {
    int retVal = 0;

    // Create adjacency list representation of the graph
    int adjacency[n][n];
    memset(adjacency, 0, (n * n * sizeof(int)));
    for (int i = 0; i < edgesSize; ++i) {
        int u = edges[i][0], v = edges[i][1];
        adjacency[u][v] = 1;
        adjacency[v][u] = 1;
    }

    bool visited[n];
    memset(visited, false, sizeof(visited));

    int component[n * n];
    int componentSize;
    int bfsQueue[n * n];
    int bfsQueueHead, bfsQueueTail, current;
    bool isComplete;
    int adjacencyNodeSize;
    for (int vertex = 0; vertex < n; vertex++) {
        if (visited[vertex] == true) {
            continue;
        }
        visited[vertex] = true;

        // BFS to find all vertices in the current component
        memset(component, 0, sizeof(component));
        componentSize = 0;
        memset(bfsQueue, 0, (n * n * sizeof(int)));
        bfsQueueHead = 0;
        bfsQueueTail = 0;
        bfsQueue[bfsQueueTail++] = vertex;
        while (bfsQueueHead < bfsQueueTail) {
            current = bfsQueue[bfsQueueHead++];
            component[componentSize++] = current;

            // Process neighbors
            for (int neighbor = 0; neighbor < n; neighbor++) {
                if (adjacency[current][neighbor] == 0) {
                    continue;
                }
                if (visited[neighbor] == true) {
                    continue;
                }
                bfsQueue[bfsQueueTail++] = neighbor;
                visited[neighbor] = true;
            }
        }

        // Check if component is complete(all vertices have the right number of edges)
        isComplete = true;
        for (int node = 0; node < componentSize; node++) {
            adjacencyNodeSize = 0;
            for (int neighbor = 0; neighbor < n; neighbor++) {
                if (adjacency[component[node]][neighbor] == 1) {
                    adjacencyNodeSize++;
                }
            }
            if (adjacencyNodeSize != componentSize - 1) {
                isComplete = false;
                break;
            }
        }
        if (isComplete == true) {
            retVal++;
        }
    }

    return retVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (int)(1e4)
    struct testCaseType {
        int n;
        int edges[MAX_SIZE][2];
        int edgesSize;
        int edgesColSize[MAX_SIZE];
    } testCase[] = {{6, {{0, 1}, {0, 2}, {1, 2}, {3, 4}}, 4, {2, 2, 2, 2}},
                    {6, {{0, 1}, {0, 2}, {1, 2}, {3, 4}, {3, 5}}, 5, {2, 2, 2, 2, 2}}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: n = 6, edges = [[0,1],[0,2],[1,2],[3,4]]
     *  Output: 3
     *
     *  Input: n = 6, edges = [[0,1],[0,2],[1,2],[3,4],[3,5]]
     *  Output: 1
     */

    int** pEdges = NULL;
    int answer = 0;
    int i, j, k;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: n = %d, edges = [", testCase[i].n);
        for (j = 0; j < testCase[i].edgesSize; ++j) {
            printf("%s[", (j == 0 ? "" : ","));
            for (k = 0; k < testCase[i].edgesColSize[j]; ++k) {
                printf("%s%d", (k == 0 ? "" : ","), testCase[i].edges[j][k]);
            }
            printf("]");
        }
        printf("]\n");

        pEdges = (int**)malloc(testCase[i].edgesSize * sizeof(int*));
        if (pEdges == NULL) {
            perror("malloc");
            return EXIT_FAILURE;
        }
        for (j = 0; j < testCase[i].edgesSize; ++j) {
            pEdges[j] = (int*)malloc(testCase[i].edgesColSize[j] * sizeof(int));
            if (pEdges[j] == NULL) {
                perror("malloc");
                for (k = 0; k < j; ++k) {
                    free(pEdges[k]);
                }
                free(pEdges);
                return EXIT_FAILURE;
            }
            memset(pEdges[j], 0, (testCase[i].edgesColSize[j] * sizeof(int)));
            memcpy(pEdges[j], testCase[i].edges[j], (testCase[i].edgesColSize[j] * sizeof(int)));
        }

        answer = countCompleteComponents(testCase[i].n, pEdges, testCase[i].edgesSize, testCase[i].edgesColSize);
        printf("Output: %d\n", answer);

        printf("\n");

        for (j = 0; j < testCase[i].edgesSize; ++j) {
            free(pEdges[j]);
            pEdges[j] = NULL;
        }
        free(pEdges);
        pEdges = NULL;
    }

    return EXIT_SUCCESS;
}
